sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -314,7 +314,17 @@ abstract class UnaryNode extends LogicalPlan {
    * expressions with the corresponding alias
    */
   protected def getAliasedConstraints(projectList: Seq[NamedExpression]): Set[Expression] = {
-    var allConstraints = child.constraints.asInstanceOf[Set[Expression]]
+    val relativeReferences = AttributeSet(projectList.collect {
+      case a: Alias => a
+    }.flatMap(_.references)) ++ outputSet
+
+    // We only care about the constraints which refer to attributes in output and aliases.
+    // For example, for a constraint 'a > b', if 'a' is aliased to 'c', we need to get aliased
+    // constraint 'c > b' only if 'b' is in output.
+    var allConstraints = child.constraints.filter { constraint =>
+      constraint.references.subsetOf(relativeReferences)
```

Comment:
Yes. You can see the benchmark in the pr description. With pruning these attributes, the running time is cut half.

If I understand your comment correctly, pruning them later in `QueryPlan` means we prune constraints which don't refer attributes in `outputSet`.

But the pruning here is happened before the pruning you pointed out, we need to reduce the constraints taken for transforming aliasing attributes to lower the computation cost.